src/Storages/StorageFile.cpp

```diff
@@ -649,6 +649,9 @@ class StorageFileSink final : public SinkToStorage
     void onFinish() override
     {
         writer->doWriteSuffix();
+        writer->flush();
+        write_buf->sync();
+        write_buf->finalize();
```

Comment:
@Avogar am I understand correctly that these hunk is a the fix (and mostly everything other is refactoring) since before it calls only `WriteBuffer::next()` which may leave some data?

BTW there is also commented `StorageFileSink::flush`, worth removing? - https://github.com/ClickHouse/ClickHouse/blob/69cd19a3422b5dd7964656f7b1a89615a7812325/src/Storages/StorageFile.cpp#L654-L657

Reply:
> @Avogar am I understand correctly that these hunk is a the fix (and mostly everything other is refactoring) since before it calls only WriteBuffer::next() which may leave some data?

Not only this change is the fix. In some WriteBuffers (for example LZMADeflatingWriteBuffer) final flush was performed only in destructor (not even in `finalize` method) and sometimes destructor could be called after query already been finished, so I moved all flushes from destructors in `finalize` method so that we can flush final data explicitly.

Reply:
> BTW there is also commented StorageFileSink::flush, worth removing? -

Yep, thanks